### PR TITLE
Energy Monitoring : Add UI helper for Tasmota integration - Update and new device

### PR DESCRIPTION
**Pull request by Terdious**
### Pull Request check-list

To ensure your Pull Request can be accepted as fast as possible, make sure to review and check all of these items:

- [x] If your changes affects code, did your write the tests?
- [x] Are tests passing? (`npm test` on both front/server)
- [x] Is the linter passing? (`npm run eslint` on both front/server)
- [x] Did you run prettier? (`npm run prettier` on both front/server)
- [x] If you are adding a new features/services, did you run integration comparator? (`npm run compare-translations` on front)
- [x] Did you add fake requests data for the demo mode (`front/src/config/demo.js`) so that the demo website is working without a backend? (if needed) See [https://demo.gladysassistant.com](https://demo.gladysassistant.com).

NOTE: these things are not required to open a PR and can be done afterwards / while the PR is open.

### Description of change

Energy Monitoring : Add UI helper for Tasmota integration

## Summary by CodeRabbit

* **New Features**
  * Smarter energy detection and automatic augmentation for discovered devices, including normalization and preservation of derived energy metrics.
  * Discovery now consults the default electric meter when available to enhance energy feature linking.
  * Device notification flow made asynchronous with safer fallback behavior.

* **Bug Fixes**
  * Added validation to skip invalid device objects.

* **Tests**
  * Large expansion of tests covering discovery, energy feature handling, notification, and edge cases.

### CODE REVIEW

**Review by Pierre-Gilles — COMMENTED** [review]
Question: j'ai peur que ce changement clash avec les intégrations qui gère déjà le suivi de l'énergie: Zigbee2mqtt par exemple ?

Comment ça rend dans ce cas là ?

**Comment by Terdious** [comment]
> Question: j'ai peur que ce changement clash avec les intégrations qui gère déjà le suivi de l'énergie: Zigbee2mqtt par exemple ?
> 
> Comment ça rend dans ce cas là ?

Alors, je n'ai pas de suivi d'énergie en Zigbee, donc je n'ai pas pu tester, mais il ne devrait pas y avoir de souci car pour déterminer si il y a un besoin de création il se base sur le energy_parent_id. As-tu moyen de tester chez toi si tu en as ? Et je demande à Mutmut en plus.

**Comment by Pierre-Gilles** [comment]
@Terdious Le « problème », c’est que cela suggère aux utilisateurs Zigbee2MQTT d’utiliser le bouton de création côté frontend, alors que pour Zigbee2MQTT il existe déjà un mécanisme automatique côté backend.

D’ailleurs, pour Tasmota, est-ce qu’on ne voudrait pas justement s’orienter vers un mécanisme similaire à celui mis en place pour Zigbee ?

Concernant l’intégration MQTT, j’ai fait le choix d’un fonctionnement manuel, car l’intégration MQTT est par nature entièrement manuelle.

Du coup, pour les intégrations dites « automatiques », est-ce qu’on ne préférerait pas privilégier une approche totalement automatique, afin d’éviter toute confusion côté utilisateur ?

**Comment by Terdious** [comment]
Mmmh, je ne suis pas sûr, cela ne suggère rien car rien ne l'indique / ne l'oblige. Les fonctionnalités de Zigbee seront toujours créées automatiquement et la vue d'édition ne proposera pas d'en créer de nouveaux.

La différence sur Tasmota (enfin visiblement je pense) est qu'il y a des index/fonctionnalités d'NRJ * 3 dont 2 ne sont pas nécessaire dans le suivi d'énergie : énergie total / énergie aujourd'hui / énergie hier. On ne souhaite pas créé pour le tout je pense

**Comment by Pierre-Gilles** [comment]
> Les fonctionnalités de Zigbee seront toujours créées automatiquement et la vue d'édition ne proposera pas d'en créer de nouveaux.

Pour les utilisateurs ayant déjà effectué la migration, oui.

En revanche, pour un utilisateur qui n’a pas encore migré, l’interface laisse entendre que la création se fait manuellement, fonctionnalité par fonctionnalité.

Dans le cas d’un Lixee TIC avec des dizaines d’index, cela devient extrêmement chronophage et l’expérience utilisateur risque d’être franchement mauvaise.
D’autant plus que nous avons déjà une expérience automatique codée, et l'utilisateur ne le verra pas du coup ! 

> La différence sur Tasmota (enfin visiblement je pense) est qu'il y a des index/fonctionnalités d'NRJ * 3 dont 2 ne sont pas nécessaire dans le suivi d'énergie : énergie total / énergie aujourd'hui / énergie hier. On ne souhaite pas créé pour le tout je pense

Justement, qu’est-ce qui empêche l’intégration automatique de gérer ce cas intelligemment ?
En ne créant que les fonctionnalités pertinentes.

Le risque avec l’approche actuelle, c’est de déplacer la responsabilité du “quoi créer” vers l’utilisateur, alors que c’est précisément quelque chose qu’il ne sait pas (et ne devrait pas avoir à savoir).

**Comment by Terdious** [comment]
Je vais reprendre le développement en ce sens.

**Comment by Terdious** [comment]
Salut Pierre-Gilles,

J’ai repris la PR en suivant tes retours et nos discussions.

**Ce qui a été fait :**
- Côté Tasmota, la création auto des fonctionnalités de conso/coût ne se fait que si une feature ENERGY:Total est détectée (pas pour Today/Yesterday).
- Le bouton “Mettre à jour” reste dans la découverte : il sert à mettre à jour les appareils existants, exactement comme Zigbee2MQTT pour ajouter les 2 fonctionnalités conso / coût.
- La découverte GET/POST est désormais alignée (même logique, même rendu).
- J’ai factorisé la logique commune dans un helper dédié et mis des tests pour couvrir tous les cas (y compris MQTT).

**Ce qui est volontairement laissé tel quel pour l’instant :**
- L’UI helper reste global (pas spécifique à Tasmota), car je pense que ça pourra servir à d’autres intégrations qui ne gèrent pas encore tout côté backend.
- Je n’ai pas ajouté d’auto-création pour Today/Yesterday (l’utilisateur peut les créer manuellement si besoin). J'en ai absolument besoin pour ma part, et ça peut être le cas d'autre personnes.

Si ça te va comme approche, tu peux review.

File: server/services/tasmota/lib/tasmota.buildDiscoveredDevice.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+      device.features.push(feature);
+    }
+  });
+};
+
+const buildDiscoveredDevice = (device, existing, defaultElectricMeterDeviceFeatureId) => {
+  ensureExistingEnergyDerivedFeatures(device, existing);
+
+  const mergedDevice = mergeDevices(device, existing);
+  const beforeFeaturesCount = Array.isArray(mergedDevice.features) ? mergedDevice.features.length : 0;
```
[the remark this answers is not in the document]
Comment (Terdious) [reply server/services/tasmota/lib/tasmota.buildDiscoveredDevice.js]: Tu as raison : j’ai été plus loin que la suggestion minimale de coderabbit. Pourquoi ?

Coderabbit proposait 2 choses :

déplacer la ré‑injection après mergeDevices,
ne ré‑injecter les features dérivées que si le parent existe.
Le “parent” est chaîné :
cost → parent = consumption → parent = feature index (ex. ENERGY:Total).
Si on applique la version minimale avec un simple parentExists basé sur mergedDevice.features, on peut avoir :

cost ajouté alors que consumption n’est pas (ou inversement),
ou ne rien ré‑injecter si la feature parent n’est pas dans la découverte, même si elle existe dans existing.
Mon correctif fait deux choses supplémentaires :

si le parent est absent du device découvert mais existe dans existing, on le ré‑injecte pour garder la chaîne intacte,
on évite d’ajouter cost si consumption n’est pas ré‑injectable (pas de parent → pas d’orphelin).
Donc oui, plus de lignes, mais ça évite une classe de bugs subtils autour des parents ou features non redécouverts.

File: server/services/energy-monitoring/utils/addEnergyFeatures.js [hunk 2]
```diff
@@ -3,12 +3,17 @@ const uuid = require('uuid');
...
 const { ENERGY_INDEX_FEATURE_TYPES } = require('./constants');
 
-const addEnergyFeatures = (device, defaultElectricMeterDeviceFeatureId) => {
+const addEnergyFeatures = (device, defaultElectricMeterDeviceFeatureId, options = {}) => {
+  if (!device || !Array.isArray(device.features)) {
+    return device;
+  }
+
+  const { filterIndexFeature } = options;
+  const hasFilter = typeof filterIndexFeature === 'function';
```
Comment (Pierre-Gilles) [inline server/services/energy-monitoring/utils/addEnergyFeatures.js]: Je suis pas fan du pattern j'avoue :) 

Je pense que c'est à l'intégration Tasmota de n'appeler cette fonction que si la fonction doit être appelée, mais ce n'est pas au service "energy-monitoring" de faire ce filtrage

File: server/services/tasmota/lib/tasmota.buildDiscoveredDevice.js [hunk 3]
```diff
@@ -0,0 +1,100 @@
+const { addEnergyFeatures } = require('../../energy-monitoring/utils/addEnergyFeatures');
+const { mergeDevices } = require('../../../utils/device');
+const { DEVICE_FEATURE_CATEGORIES, DEVICE_FEATURE_TYPES } = require('../../../utils/constants');
+
+const isTasmotaTotalEnergyIndex = (feature) => {
+  const externalId = feature.external_id || feature.selector || '';
+  return (
+    feature.category === DEVICE_FEATURE_CATEGORIES.ENERGY_SENSOR &&
+    feature.type === DEVICE_FEATURE_TYPES.ENERGY_SENSOR.ENERGY &&
+    (/:ENERGY:Total(\d+)?$/i.test(externalId) || /StatusSNS\.ENERGY\.Total$/i.test(externalId))
```
Comment (Pierre-Gilles) [inline server/services/tasmota/lib/tasmota.buildDiscoveredDevice.js]: Ce bout de code doit être dans avant l'appel à la fonction addEnergyFeatures()

**Review by Pierre-Gilles — CHANGES_REQUESTED** [review]

File: server/services/energy-monitoring/utils/addEnergyFeatures.js [hunk 4]
```diff
@@ -20,13 +23,27 @@ const addEnergyFeatures = (device, defaultElectricMeterDeviceFeatureId) => {
...
+      if (f.energy_parent_id && f.energy_parent_id === indexFeature.id) {
+        return true;
+      }
+
+      if (!f.external_id || !indexFeature.external_id) {
+        return false;
+      }
+
+      return (
+        f.external_id === `${indexFeature.external_id}_consumption` ||
```
Comment (Pierre-Gilles) [inline server/services/energy-monitoring/utils/addEnergyFeatures.js]: Can't you set the energy_parent_id properly in the Tasmota integration?

File: server/services/energy-monitoring/utils/addEnergyFeatures.js [hunk 5]
```diff
@@ -51,13 +68,26 @@ const addEnergyFeatures = (device, defaultElectricMeterDeviceFeatureId) => {
...
+      if (f.energy_parent_id && f.energy_parent_id === consumptionFeatureId) {
+        return true;
+      }
+
+      if (!f.external_id || !indexFeature.external_id) {
+        return false;
+      }
+
+      return (
+        f.external_id === `${indexFeature.external_id}_cost` || f.external_id === `${indexFeature.external_id}:cost`
```
Comment (Pierre-Gilles) [inline server/services/energy-monitoring/utils/addEnergyFeatures.js]: Same

**Review by Pierre-Gilles — CHANGES_REQUESTED** [review]

File: server/services/energy-monitoring/utils/addEnergyFeatures.js (hunk 4 above)
Comment (Terdious) [reply server/services/energy-monitoring/utils/addEnergyFeatures.js]: I added an external_id fallback to avoid re‑creating consumption/cost features if the user re‑parented them in ‘Energy Monitoring’. But since Tasmota/Zigbee2mqtt go through mergeDevices, we don’t duplicate anything in the DB. So we can revert addEnergyFeatures to its original version and fix the Tasmota integration to set energy_parent_id properly.

File: server/services/energy-monitoring/utils/addEnergyFeatures.js (hunk 5 above)
Comment (Terdious) [reply server/services/energy-monitoring/utils/addEnergyFeatures.js]: Done

File: server/test/services/tasmota/lib/mqtt/handle-message/device-creation/tasmota.mqtt.handleMessage.deviceCreation-energy-feature.test.js [hunk 6]
```diff
@@ -110,11 +113,14 @@ describe('Tasmota - MQTT - create device with ENERGY features', () => {
     assert.calledWith(mqttService.device.publish, 'cmnd/tasmota-device-topic/STATUS', '8');
   });
 
-  it('decode STATUS8 message', () => {
+  it('decode STATUS8 message', async () => {
     tasmotaHandler.handleMessage('stat/tasmota-device-topic/STATUS', JSON.stringify(messages.STATUS));
     tasmotaHandler.handleMessage('stat/tasmota-device-topic/STATUS5', JSON.stringify(messages.STATUS5));
     mqttService.device.publish.resetHistory();
     tasmotaHandler.handleMessage('stat/tasmota-device-topic/STATUS8', JSON.stringify(messages.STATUS8));
+    await new Promise((resolve) => {
```
Comment (Pierre-Gilles) [inline server/test/services/tasmota/lib/mqtt/handle-message/device-creation/tasmota.mqtt.handleMessage.deviceCreation-energy-feature.test.js]: This is not the solution, the function notifyNewDevice is now async and should be called with await in the handleMessage function (currently, it's not the case)

File: server/services/tasmota/lib/tasmota.notifyNewDevice.js [hunk 7]
```diff
@@ -7,12 +10,43 @@ const { EVENTS } = require('../../../utils/constants');
  * @example
  * notifyNewDevice(discorveredDevice)
  */
-function notifyNewDevice(device, event) {
-  const payload = this.mergeWithExistingDevice(device);
-  this.gladys.event.emit(EVENTS.WEBSOCKET.SEND_ALL, {
-    type: event,
-    payload,
-  });
+async function notifyNewDevice(device, event) {
```
Comment (Pierre-Gilles) [inline server/services/tasmota/lib/tasmota.notifyNewDevice.js:14]: This function is now async, and should be called asynchronously in places where it's used

**Review by Pierre-Gilles — CHANGES_REQUESTED** [review]

File: server/services/tasmota/lib/tasmota.notifyNewDevice.js (hunk 7 above)
Comment (Terdious) [reply server/services/tasmota/lib/tasmota.notifyNewDevice.js:14]: Thanks for the review. 
I’ve now fixed all call sites (MQTT + HTTP) and updated every impacted test to await handleMessage/notifyNewDevice properly. The whole set of Tasmota device‑creation tests has been adjusted accordingly.

I missed that notifyNewDevice had become async and didn’t propagate the await everywhere. I should have double‑checked all call sites and tests before pushing.

File: server/test/services/tasmota/lib/mqtt/handle-message/device-creation/tasmota.mqtt.handleMessage.deviceCreation-energy-feature.test.js (hunk 6 above)
Comment (Terdious) [reply server/test/services/tasmota/lib/mqtt/handle-message/device-creation/tasmota.mqtt.handleMessage.deviceCreation-energy-feature.test.js]: Fixed: handleMessage now awaits notifyNewDevice and all STATUS8 tests were updated with await.

**Review by Pierre-Gilles — APPROVED** [review]
Good for me! Thanks for this PR 🙌
